<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>singlepp: C++ port of SingleR</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  extensions: ["tex2jax.js"],
  jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">singlepp
   </div>
   <div id="projectbrief">A C++ library for cell type classification</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.8 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">C++ port of SingleR </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="md__2github_2workspace_2README"></a> <img src="https://github.com/singler-inc/singlepp/actions/workflows/run-tests.yaml/badge.svg" alt="Unit tests" style="pointer-events: none;" class="inline"/> <img src="https://github.com/singler-inc/singlepp/actions/workflows/doxygenate.yaml/badge.svg" alt="Documentation" style="pointer-events: none;" class="inline"/> <img src="https://github.com/singler-inc/singlepp/actions/workflows/compare-r.yaml/badge.svg" alt="R comparison" style="pointer-events: none;" class="inline"/> <a href="https://codecov.io/gh/singler-inc/singlepp"><img src="https://codecov.io/gh/singler-inc/singlepp/branch/master/graph/badge.svg?token=OYTGM9IRSE" alt="codecov" style="pointer-events: none;" class="inline"/></a></p>
<h1>Overview</h1>
<p>This repository contains a C++ port of the <a href="https://bioconductor.org/packages/SingleR"><b>SingleR</b></a> R package for automated cell type annotation. Given a test matrix of single-cell (expression) values, it compares each cell to a reference dataset with known cell type labels. Scoring is based on Spearman's rank correlation across the marker genes for each labels, with additional fine-tuning to distinguish between closely related labels. <b>singlepp</b> returns these scores along with the best label for each cell in the test dataset. We provide methods for annotation based on a single reference as well as integration of labels across multiple references.</p>
<h1>Quick start</h1>
<p><b>singlepp</b> is a header-only library, so it can be easily used by just <code>#include</code>ing the relevant source files. Assuming the reference matrix, labels and markers are available, we can easily run the classification:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &quot;<a class="code" href="singlepp_8hpp.html">singlepp/singlepp.hpp</a>&quot;</span></div>
<div class="line"> </div>
<div class="line"><span class="comment">// Prepare the reference matrix as a tatami::NumericMatrix.</span></div>
<div class="line">ref_mat;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Prepare a vector of labels, one per column of ref_mat.</span></div>
<div class="line">ref_labels;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Prepare a vector of vectors of markers for pairwise comparisons between labels.</span></div>
<div class="line">ref_markers;</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Building the classifier.</span></div>
<div class="line"><a class="code hl_typedef" href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::TrainSingleOptions&lt;int, double&gt;</a> train_opt;</div>
<div class="line"><span class="keyword">auto</span> trained = <a class="code hl_function" href="namespacesinglepp.html#abcea821671f9339e4e933be2ed123c6e">singlepp::train_single</a>(ref_mat, ref_labels.data(), train_opt);</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Running the classification on the test matrix.</span></div>
<div class="line"><a class="code hl_typedef" href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::ClassifySingleOptions&lt;double&gt;</a> class_opt;</div>
<div class="line"><span class="keyword">auto</span> res = <a class="code hl_typedef" href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::classify_single&lt;int&gt;</a>(test_mat, trained, class_opt);</div>
<div class="ttc" id="anamespacesinglepp_html_a87e4be0939142d883ad98944e6336a5c"><div class="ttname"><a href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::Markers</a></div><div class="ttdeci">std::vector&lt; std::vector&lt; std::vector&lt; Index_ &gt; &gt; &gt; Markers</div><div class="ttdef"><b>Definition</b> Markers.hpp:40</div></div>
<div class="ttc" id="anamespacesinglepp_html_abcea821671f9339e4e933be2ed123c6e"><div class="ttname"><a href="namespacesinglepp.html#abcea821671f9339e4e933be2ed123c6e">singlepp::train_single</a></div><div class="ttdeci">TrainedSingle&lt; Index_, Float_ &gt; train_single(const tatami::Matrix&lt; Value_, Index_ &gt; &amp;ref, const Label_ *labels, Markers&lt; Index_ &gt; markers, const TrainSingleOptions&lt; Index_, Float_ &gt; &amp;options)</div><div class="ttdef"><b>Definition</b> train_single.hpp:180</div></div>
<div class="ttc" id="asinglepp_8hpp_html"><div class="ttname"><a href="singlepp_8hpp.html">singlepp.hpp</a></div><div class="ttdoc">Umbrella header for the singlepp library.</div></div>
</div><!-- fragment --><p>See the <a href="https://singler-inc.github.io/singlepp">reference documentation</a> for more details.</p>
<h1>Identifying markers</h1>
<p>Given a reference dataset, <b>singlepp</b> implements a simple method of identifying marker genes between labels. This is based on ranking the differences in median log-expression values between labels and is the "classic" method provided in the original <b>SingleR</b> package.</p>
<div class="fragment"><div class="line"><a class="code hl_struct" href="structsinglepp_1_1ChooseClassicMarkersOptions.html">singlepp::ChooseClassicMarkersOptions</a> mrk;</div>
<div class="line"><span class="keyword">auto</span> classic_markers = <a class="code hl_typedef" href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::choose_classic_markers&lt;int&gt;</a>(</div>
<div class="line">    ref_mat.get(),</div>
<div class="line">    ref_labels.data(),</div>
<div class="line">    m_opt</div>
<div class="line">);</div>
<div class="ttc" id="astructsinglepp_1_1ChooseClassicMarkersOptions_html"><div class="ttname"><a href="structsinglepp_1_1ChooseClassicMarkersOptions.html">singlepp::ChooseClassicMarkersOptions</a></div><div class="ttdoc">Options for choose_classic_markers().</div><div class="ttdef"><b>Definition</b> choose_classic_markers.hpp:40</div></div>
</div><!-- fragment --><p>The <code>classic_markers</code> can then be directly used in <code>train_single()</code>. Of course, other marker detection schemes can be used, depending on the type of reference dataset. For single-cell references, users may be interested in some of the differential analysis methods in the <a href="https://github.com/libscran/scran_markers"><b>libscran</b></a> library.</p>
<p>By default, it is expected that the <code>markers</code> supplied to <code>train_single()</code> has already been filtered to only the top markers for each pairwise comparison. However, in some cases, it might be more convenient for <code>markers</code> to contain a ranking of all genes such that the desired subset of top markers can be chosen later. This is achieved by setting <code>TrainSingleOptions::top</code> to the desired number of markers per comparison, e.g., for 20 markers:</p>
<div class="fragment"><div class="line">train_opt.<a class="code hl_variable" href="structsinglepp_1_1TrainSingleOptions.html#aa72f73c9937301bd90c2fec19f523479">top</a> = 20;</div>
<div class="line"><span class="keyword">auto</span> trained20 = <a class="code hl_function" href="namespacesinglepp.html#abcea821671f9339e4e933be2ed123c6e">singlepp::train_single</a>(</div>
<div class="line">    ref_mat,</div>
<div class="line">    ref_labels.data(),</div>
<div class="line">    train_opt</div>
<div class="line">);</div>
<div class="ttc" id="astructsinglepp_1_1TrainSingleOptions_html_aa72f73c9937301bd90c2fec19f523479"><div class="ttname"><a href="structsinglepp_1_1TrainSingleOptions.html#aa72f73c9937301bd90c2fec19f523479">singlepp::TrainSingleOptions::top</a></div><div class="ttdeci">int top</div><div class="ttdef"><b>Definition</b> train_single.hpp:36</div></div>
</div><!-- fragment --><p>Doing so is roughly equivalent to slicing each vector in <code>markers</code> to the top 20 entries before calling <code>train_single()</code>. In fact, calling <code>set_top()</code> is the better approach when intersecting feature spaces - see below - as the top set will not be contaminated by genes that are not present in the test dataset.</p>
<h1>Intersecting feature sets</h1>
<p>Often the reference dataset will not have the same genes as the test dataset. To handle this case, users should use <code>train_single_intersect()</code> with identifiers for the rows of the reference and test matrices.</p>
<div class="fragment"><div class="line">test_names; <span class="comment">// vector of feature IDs for the test data</span></div>
<div class="line">ref_names; <span class="comment">// vector of feature IDs for the reference data</span></div>
<div class="line"> </div>
<div class="line"><span class="keyword">auto</span> trained_intersect = <a class="code hl_function" href="namespacesinglepp.html#a606d38abb5608efdfdb0c3bdb945ef23">singlepp::train_single_intersect</a>(</div>
<div class="line">    test_mat.nrow(),</div>
<div class="line">    test_names.data(),</div>
<div class="line">    ref_mat,</div>
<div class="line">    ref_names.data(),</div>
<div class="line">    ref_labels.data(), </div>
<div class="line">    ref_markers,</div>
<div class="line">    train_opt</div>
<div class="line">);</div>
<div class="ttc" id="anamespacesinglepp_html_a606d38abb5608efdfdb0c3bdb945ef23"><div class="ttname"><a href="namespacesinglepp.html#a606d38abb5608efdfdb0c3bdb945ef23">singlepp::train_single_intersect</a></div><div class="ttdeci">TrainedSingleIntersect&lt; Index_, Float_ &gt; train_single_intersect(const Intersection&lt; Index_ &gt; &amp;intersection, const tatami::Matrix&lt; Value_, Index_ &gt; &amp;ref, const Label_ *labels, Markers&lt; Index_ &gt; markers, const TrainSingleOptions&lt; Index_, Float_ &gt; &amp;options)</div><div class="ttdef"><b>Definition</b> train_single.hpp:308</div></div>
</div><!-- fragment --><p>Then, <code>classify_single_intersect()</code> will perform classification using only the intersection of genes:</p>
<div class="fragment"><div class="line"><span class="keyword">auto</span> res_intersect = <a class="code hl_typedef" href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::classify_single_intersect&lt;int&gt;</a>(</div>
<div class="line">    test_mat,</div>
<div class="line">    trained_intersect,</div>
<div class="line">    class_opt</div>
<div class="line">);</div>
</div><!-- fragment --><p>The gene identifiers can be anything that can be hashed and compared. These are most commonly <code>std::string</code>s but can also be integers (e.g., for Entrez IDs).</p>
<h1>Integrating results across references</h1>
<p>To combine results from multiple references, we first need to perform classification within each reference. Let's say we have two references A and B:</p>
<div class="fragment"><div class="line"><span class="keyword">auto</span> trainA = <a class="code hl_function" href="namespacesinglepp.html#abcea821671f9339e4e933be2ed123c6e">singlepp::train_single</a>(refA_mat, refA_labels.data(), refA_markers, train_opt);</div>
<div class="line"><span class="keyword">auto</span> resA = <a class="code hl_typedef" href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::classify_single&lt;int&gt;</a>(test_mat, trainA, class_opt);</div>
<div class="line"> </div>
<div class="line"><span class="keyword">auto</span> trainB = <a class="code hl_function" href="namespacesinglepp.html#abcea821671f9339e4e933be2ed123c6e">singlepp::train_single</a>(refB_mat, refB_labels.data(), refB_markers, train_opt);</div>
<div class="line"><span class="keyword">auto</span> resB = <a class="code hl_typedef" href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::classify_single&lt;int&gt;</a>(test_mat, trainB, class_opt);</div>
</div><!-- fragment --><p>We build the integrated classifier:</p>
<div class="fragment"><div class="line">std::vector&lt;singlepp::TrainIntegratedInput&lt;double, int, int&gt; &gt; inputs;</div>
<div class="line">inputs.push_back(<a class="code hl_function" href="namespacesinglepp.html#a280f3bfb8b5a2e7e51233bc7382a8db1">singlepp::prepare_integrated_input</a>(refA_mat, refA_labels.data(), preA));</div>
<div class="line">inputs.push_back(<a class="code hl_function" href="namespacesinglepp.html#a280f3bfb8b5a2e7e51233bc7382a8db1">singlepp::prepare_integrated_input</a>(refB_mat, refB_labels.data(), preB));</div>
<div class="line"> </div>
<div class="line"><a class="code hl_struct" href="structsinglepp_1_1TrainIntegratedOptions.html">singlepp::TrainIntegratedOptions</a> ti_opt;</div>
<div class="line"><span class="keyword">auto</span> train_integrated = <a class="code hl_function" href="namespacesinglepp.html#af198ad22e020701691c9bb8326401a46">singlepp::train_integrated</a>(inputs, ti_opt);</div>
<div class="ttc" id="anamespacesinglepp_html_a280f3bfb8b5a2e7e51233bc7382a8db1"><div class="ttname"><a href="namespacesinglepp.html#a280f3bfb8b5a2e7e51233bc7382a8db1">singlepp::prepare_integrated_input</a></div><div class="ttdeci">TrainIntegratedInput&lt; Value_, Index_, Label_ &gt; prepare_integrated_input(const tatami::Matrix&lt; Value_, Index_ &gt; &amp;ref, const Label_ *labels, const TrainedSingle&lt; Index_, Float_ &gt; &amp;trained)</div><div class="ttdef"><b>Definition</b> train_integrated.hpp:73</div></div>
<div class="ttc" id="anamespacesinglepp_html_af198ad22e020701691c9bb8326401a46"><div class="ttname"><a href="namespacesinglepp.html#af198ad22e020701691c9bb8326401a46">singlepp::train_integrated</a></div><div class="ttdeci">TrainedIntegrated&lt; Index_ &gt; train_integrated(const std::vector&lt; TrainIntegratedInput&lt; Value_, Index_, Label_ &gt; &gt; &amp;inputs, const TrainIntegratedOptions &amp;options)</div><div class="ttdef"><b>Definition</b> train_integrated.hpp:450</div></div>
<div class="ttc" id="astructsinglepp_1_1TrainIntegratedOptions_html"><div class="ttname"><a href="structsinglepp_1_1TrainIntegratedOptions.html">singlepp::TrainIntegratedOptions</a></div><div class="ttdoc">Options for train_integrated().</div><div class="ttdef"><b>Definition</b> train_integrated.hpp:263</div></div>
</div><!-- fragment --><p>And then we can finally run the scoring. For each cell in the test dataset, <code>classify_integrated()</code> picks the best label among the assignments from each individual reference.</p>
<div class="fragment"><div class="line"><a class="code hl_typedef" href="namespacesinglepp.html#a87e4be0939142d883ad98944e6336a5c">singlepp::ClassifyIntegratedOptions&lt;double&gt;</a> ci_opt;</div>
<div class="line"><span class="keyword">auto</span> ires = single.run(test_mat, train_integrated, ci_opt);</div>
<div class="line">ires.best; <span class="comment">// index of the best reference.</span></div>
</div><!-- fragment --><h1>Building projects</h1>
<h2>CMake with <code>FetchContent</code></h2>
<p>If you're using CMake, you just need to add something like this to your <code>CMakeLists.txt</code>:</p>
<div class="fragment"><div class="line">include(FetchContent)</div>
<div class="line"> </div>
<div class="line">FetchContent_Declare(</div>
<div class="line">  singlepp</div>
<div class="line">  GIT_REPOSITORY https://github.com/singler-inc/singlepp</div>
<div class="line">  GIT_TAG master # or any version of interest</div>
<div class="line">)</div>
<div class="line"> </div>
<div class="line">FetchContent_MakeAvailable(singlepp)</div>
</div><!-- fragment --><p>Then you can link to <b>singlepp</b> to make the headers available during compilation:</p>
<div class="fragment"><div class="line"># For executables:</div>
<div class="line">target_link_libraries(myexe singlepp)</div>
<div class="line"> </div>
<div class="line"># For libaries</div>
<div class="line">target_link_libraries(mylib INTERFACE singlepp)</div>
</div><!-- fragment --><h2>CMake with <code>find_package()</code></h2>
<div class="fragment"><div class="line">find_package(singler_singlepp CONFIG REQUIRED)</div>
<div class="line">target_link_libraries(mylib INTERFACE singler::singlepp)</div>
</div><!-- fragment --><p>To install the library, use:</p>
<div class="fragment"><div class="line">mkdir build &amp;&amp; cd build</div>
<div class="line">cmake .. -DSINGLEPP_TESTS=OFF</div>
<div class="line">cmake --build . --target install</div>
</div><!-- fragment --><p>By default, this will use <code>FetchContent</code> to fetch all external dependencies. If you want to install them manually, use <code>-DSINGLEPP_FETCH_EXTERN=OFF</code>. See the tags in <a href="extern/CMakeLists.txt"><code>extern/CMakeLists.txt</code></a> to find compatible versions of each dependency.</p>
<h2>Manual</h2>
<p>If you're not using CMake, the simple approach is to just copy the files in <code>include/</code> - either directly or with Git submodules - and include their path during compilation with, e.g., GCC's <code>-I</code>. This requires the external dependencies listed in <a href="extern/CMakeLists.txt"><code>extern/CMakeLists.txt</code></a>, which also need to be made available during compilation.</p>
<h1>References</h1>
<p>Aran D et al. (2019). Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage. <em>Nat. Immunol.</em> 20, 163-172 </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.8
</small></address>
</body>
</html>
